### PR TITLE
AN-880 SSH Tunnel Revision

### LINKED ISSUE #79: [BUG] SSH Tunnel Needs rework
Currently the SSH tunnel has some issues and is poorly documented. It was removed due to this problem, we need to fix it and revive the documentation.

Some details:
When strict host key checking is enabled, there is a weird error

On Tableau Desktop, when i enable strict host key checking but did not modify known hosts file yet, it fails with weird error message. Log shows problem is related to host key checking but error message on Tableau is about not adding driver.

Tableau crashes when strict host key check is disabled.

Also at random times when using SSH tunneling and disabling the strict host ket check causes the tableau to freeze.

SSH Tunnel Docs unclear

The discussion of SSH tunnels is really unclear. If I am running over an SSH tunnel do I also need to give all these parameters to the driver? The sample code in the docs just hangs for me while trying to connect over an SSH tunnel so the docs need to make very explicit when these parameters are needed and why. Having to put pem file info etc into text inside code is not secure.

Establishing a connection manually via a URL could use a concrete example walkthrough when configuring an SSH tunnel.

### PR DESCRIPTION
**Pull request by xiazcy**
### Summary

AN-880 SSH Tunnel Revision

### Description

Updated the SSH tunnel taco and documentation for Tableau, strict host key checking should be working properly with connection related exception instead of "driver not found" exception on Tableau upon errors. Refactored Neptune Driver class.

### Related Issue

https://bitquill.atlassian.net/browse/AN-880

### Additional Reviewers
@birschick-bq
@lyndonb-bq
@xiazcy
@simonz-bq
@alexey-temnikov
@kylepbit

### CODE REVIEW

File: tableau-connector/src/connectionProperties.js [hunk 1]
```diff
@@ -27,12 +33,29 @@
...
     for (var key in params) {
+
         formattedParams.push(connectionHelper.formatKeyValuePair(key, params[key]));
     }
 
-    logging.log("formattedParams=" + formattedParams);
+    // Format the additional option 'keys=value' pairs into the connection properties.
+    if (attr['v-additional-properties']) {
+        var additionalProperties = attr['v-additional-properties'].split(";");
+        for (var i = 0; i < additionalProperties.length; i++) {
```
Comment (simonz-bq) [inline tableau-connector/src/connectionProperties.js]: Using `let` is more appropriate here.

File: src/main/java/software/aws/neptune/NeptuneDriver.java [hunk 2]
```diff
@@ -76,62 +68,75 @@ public boolean acceptsURL(final String url) throws SQLException {
...
-        if (GREMLIN.equalsIgnoreCase(language) || SQL_GREMLIN.equalsIgnoreCase(language)) {
-            return new GremlinConnectionProperties(properties);
-        }
-        if (SPARQL.equalsIgnoreCase(language)) {
-            return new SparqlConnectionProperties(properties);
-        }
-        return new OpenCypherConnectionProperties(properties);
-    }
+    private class DriverConnectionFactory {
+        private final Pattern connStringPattern = Pattern.compile(CONN_STRING_PREFIX + "(\\w+)://(.*)");
```
Comment (simonz-bq) [inline src/main/java/software/aws/neptune/NeptuneDriver.java:84]: Not really a problem nor is it necessarily within the scope of this ticket, but IMO regex should be a last resort.

It seems like we can just use a `String.contains` or something along those lines to determine the connection?

Either way it doesn't need to be addressed here.

File: src/main/java/software/aws/neptune/NeptuneDriver.java [hunk 3]
```diff
@@ -76,62 +68,75 @@ public boolean acceptsURL(final String url) throws SQLException {
...
+        public java.sql.Connection getConnection(final String url, final Properties info) throws SQLException {
+            final String language = getLanguage(url, connStringPattern);
+            final String propertyString = getPropertyString(url, connStringPattern);
+            final String firstPropertyKey = getFirstPropertyKey(language);
+            final Properties properties = setProperties(propertyString, firstPropertyKey, info);
+
+            if (OPENCYPHER.equalsIgnoreCase(language)) {
+                return new OpenCypherConnection(new OpenCypherConnectionProperties(properties));
+            }
+            if (GREMLIN.equalsIgnoreCase(language)) {
```
Comment (simonz-bq) [inline src/main/java/software/aws/neptune/NeptuneDriver.java]: A `switch` here might be more effective.

File: src/main/java/software/aws/neptune/NeptuneDriver.java [hunk 4]
```diff
@@ -76,62 +68,75 @@ public boolean acceptsURL(final String url) throws SQLException {
...
+            }
+            LOGGER.error("Encountered Neptune JDBC connection string but failed to parse driver language.");
+            throw SqlError.createSQLException(
+                    LOGGER, SqlState.CONNECTION_EXCEPTION,
+                    SqlError.INVALID_CONNECTION_PROPERTY);
         }
-        if (GREMLIN.equalsIgnoreCase(language) || SQL_GREMLIN.equalsIgnoreCase(language)) {
-            return GremlinConnectionProperties.CONTACT_POINT_KEY;
+
+        private Properties setProperties(final String propertyString, final String firstPropertyKey,
```
Comment (simonz-bq) [inline src/main/java/software/aws/neptune/NeptuneDriver.java]: I could be lacking context here, but is the naming of this correct? It sounds like a setter but is returning the Properties as well.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: src/main/java/software/aws/neptune/NeptuneDriver.java (hunk 4 above)
Comment (xiazcy) [reply src/main/java/software/aws/neptune/NeptuneDriver.java:114]: It sets and returns the property, just for the class though. Is there a better convention to follow?

File: src/main/java/software/aws/neptune/NeptuneDriver.java (hunk 2 above)
Comment (xiazcy) [reply src/main/java/software/aws/neptune/NeptuneDriver.java:84]: This was taken from existing implementation, and I thought it was used as a last resort? @lyndonb-bq did you try others?

File: src/main/java/software/aws/neptune/NeptuneDriver.java (hunk 4 above)
Comment (xiazcy) [reply src/main/java/software/aws/neptune/NeptuneDriver.java]: It takes properties from string, and makes them into a property to be returned. It's just a helper for the above function. Is there a different naming convention I should use to make it more clear?

File: src/main/java/software/aws/neptune/NeptuneDriver.java (hunk 3 above)
Comment (xiazcy) [reply src/main/java/software/aws/neptune/NeptuneDriver.java]: I'll refactor into a `switch`

File: tableau-connector/src/connectionProperties.js (hunk 1 above)
Comment (xiazcy) [reply tableau-connector/src/connectionProperties.js]: Noted.

**Review by simonz-bq — APPROVED** [review]

File: src/main/java/software/aws/neptune/NeptuneDriver.java [hunk 5]
```diff
@@ -76,62 +68,74 @@ public boolean acceptsURL(final String url) throws SQLException {
...
+            }
+            LOGGER.error("Encountered Neptune JDBC connection string but failed to parse endpoint property.");
+            throw SqlError.createSQLException(
+                    LOGGER, SqlState.CONNECTION_EXCEPTION,
+                    SqlError.INVALID_CONNECTION_PROPERTY);
         }
-        return OpenCypherConnectionProperties.ENDPOINT_KEY;
+
     }
+
```
Comment (lyndonbauto) [inline src/main/java/software/aws/neptune/NeptuneDriver.java]:
```suggestion
```

File: src/main/java/software/aws/neptune/NeptuneDriver.java [hunk 6]
```diff
@@ -76,62 +68,74 @@ public boolean acceptsURL(final String url) throws SQLException {
...
+            if (SPARQL.equalsIgnoreCase(language)) {
+                return SparqlConnectionProperties.ENDPOINT_KEY;
+            }
+            LOGGER.error("Encountered Neptune JDBC connection string but failed to parse endpoint property.");
+            throw SqlError.createSQLException(
+                    LOGGER, SqlState.CONNECTION_EXCEPTION,
+                    SqlError.INVALID_CONNECTION_PROPERTY);
         }
-        return OpenCypherConnectionProperties.ENDPOINT_KEY;
+
```
Comment (lyndonbauto) [inline src/main/java/software/aws/neptune/NeptuneDriver.java]:
```suggestion
```

**Review by lyndonbauto — APPROVED** [review]

**Pull request merged**